packages/react-dom/src/test-utils/ReactTestUtilsAct.js

```diff
@@ -0,0 +1,219 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {Thenable} from 'react-reconciler/src/ReactFiberScheduler';
+
+import warningWithoutStack from 'shared/warningWithoutStack';
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+// Keep in sync with ReactDOMUnstableNativeDependencies.js
+// ReactDOM.js, and ReactTestUtils.js:
+const [
+  /* eslint-disable no-unused-vars */
+  getInstanceFromNode,
+  getNodeFromInstance,
+  getFiberCurrentPropsFromNode,
+  injectEventPluginsByName,
+  eventNameDispatchConfigs,
+  accumulateTwoPhaseDispatches,
+  accumulateDirectDispatches,
+  enqueueStateRestore,
+  restoreStateIfNeeded,
+  dispatchEvent,
+  runEventsInBatch,
+  /* eslint-enable no-unused-vars */
+  flushPassiveEffects,
+] = ReactDOM.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.Events;
+
+const RendererHelpers = {
+  flushPassiveEffects,
+  batchedUpdates: ReactDOM.unstable_batchedUpdates,
+};
+
+const {
+  isActingUpdates,
+} = React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
+
+// the rest of this file should be exactly the same in ReactTestUtilsAct.js,
+// ReactTestRendererAct.js, ReactNoopAct.js, ReactNoopPersistentAct.js
+
+let actingUpdatesScopeDepth = 0;
+
+function incrementActingScopeDepth() {
+  actingUpdatesScopeDepth++;
+  isActingUpdates[0] = true;
+}
+
+function decrementActingScopeDepth() {
+  actingUpdatesScopeDepth--;
+  if (actingUpdatesScopeDepth === 0) {
+    isActingUpdates[0] = false;
+  }
+}
+
+let didWarnAboutMessageChannel = false;
+
+let enqueueTask;
+try {
+  // assuming we're in node, let's try to get node's
+  // version of setImmediate, bypassing fake timers if any
+  let r = require; // trick packagers not to bundle this stuff.
+  enqueueTask = r('timers').setImmediate;
+} catch (_err) {
+  // we're in a browser
+  // we can't use regular timers because they may still be faked
+  // so we try MessageChannel+postMessage instead
+  enqueueTask = function(callback) {
+    if (__DEV__) {
+      if (didWarnAboutMessageChannel === false) {
+        didWarnAboutMessageChannel = true;
+        warningWithoutStack(
+          typeof MessageChannel !== 'undefined',
+          'This browser does not have a MessageChannel implementation, ' +
+            'so enqueuing tasks via await act(async () => ...) will fail. ' +
+            'Please file an issue at https://github.com/facebook/react/issues ' +
+            'if you encounter this warning.',
+        );
+      }
+    }
+    const channel = new MessageChannel();
+    channel.port1.onmessage = callback;
+    channel.port2.postMessage(undefined);
+  };
+}
+
+function createActedUpdatesScope(callback: (onDone: (?Error) => void) => void) {
```

Comment:
What's the purpose of this closure? It looks like you're trying to support multiple, concurrent `act` sequences. I don't think we care to support that. And it wouldn't work as-is, anyway, because of the shared mutable state that lives on ReactSecretInternals.

Reply:
Oh, is the closure merely so you can fire the "overlapping act() calls" warning below? You can use a module level variable to track that, instead of this wrapper.

Reply:
the opposite, rather. because we don't want to support concurrent act sequences, we track the depth count changes, so we can warn if they do try to do so (eg- `act(async...); act(async...)` without `await`s, should warn)